File: Striver Important Interview Questions/Day 23 - Graphs/number-of-islands.cpp

// https://leetcode.com/problems/number-of-islands/description/

class Solution {
public:

    bool validPoints(int x, int y, int numRows, int numCols){
        return ( x >= 0 && x < numRows && y >= 0 && y < numCols);
    }

    void dfs(vector<vector<char>>&grid, int numRows, int numCols, vector<vector<bool>>&visited, int x, int y){

        visited[x][y] = true;

        int dx[] = {-1, 0, 1, 0};
        int dy[] = {0, 1, 0, -1};

        for(int i = 0; i < 4; i++){

            int newX = x + dx[i];
            int newY = y + dy[i];

            if(validPoints(newX, newY, numRows, numCols) && !visited[newX][newY] && grid[newX][newY] == '1'){
                dfs(grid, numRows, numCols, visited, newX, newY);
            }
        }

        return;
    }

    int numIslands(vector<vector<char>>& grid) {
        
        int numIslands = 0;
        int numRows = grid.size();
        if(numRows == 0) return numIslands;
        int numCols = grid[0].size();

        vector<vector<bool>> visited(numRows, vector<bool>(numCols, false));
        
        for(int i = 0; i < numRows; i++){
            for(int j = 0; j < numCols; j++){
                if(!visited[i][j] && grid[i][j] == '1'){
                    dfs(grid, numRows, numCols, visited, i, j);
                    numIslands++;
                }
            }
        }

        return numIslands;
    }
};

File: Neetcode-150-js/Graphs/number-of-islands.js

// https://leetcode.com/problems/number-of-islands/

/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    let numRows = grid.length;
    let ans = 0;
    if(numRows === 0) return ans;
    let numCols = grid[0].length;

    const visited = Array.from({length : numRows}, () => Array(numCols).fill(false));

    const dfs = (grid, x, y) => {
        if(x < 0 || x >= numRows || y < 0 || y >= numCols) return;

        visited[x][y] = true;
        const Dx = [-1, 0, 1, 0];
        const Dy = [0, 1, 0, -1];
        for(let i = 0; i < 4; i++){
            let newX = x + Dx[i];
            let newY = y + Dy[i];
            if((newX >= 0 && newX < numRows && newY >= 0 && newY < numCols) && !visited[newX][newY] && grid[newX][newY] === "1"){
                dfs(grid, newX, newY);
            }
        }
    }

    for(let i = 0; i < numRows; i++){
        for(let j = 0; j < numCols; j++){
            if(grid[i][j] === "1" && !visited[i][j]){
                dfs(grid, i, j)
                ans++;
            }
        }
    }

    return ans;
};